02cpp1/sec02PassByValueOrReference.md

```diff
@@ -0,0 +1,101 @@
+
+## Pass by Value
+
+Pass by Value means that we copy the value of the variable we want to pass into the function, and the function works on this copy and leaves the original alone. Any changes that the function makes to the variable will not affect the value of that variable once you leave the function's scope. To pass by value you simply write the type and name of the variable in the function parameters in the usual way. 
+```cpp
+int add(int a, int b)
+{
+    return (a + b);
+}
+```
+This function can be safer, but is not time or memory efficient if variables are complex or large in size. You should only use pass by value for large pieces of data if you need an explicit copy made to work on locally in the function body. 
+
+## Pass by Reference 
+
+Pass by Reference means that we tell the function where the original variable has been stored in memory, and we allow the function to work directly with that original variable. This has two major consequences:
+   * We only pass a memory address -- 4 bytes, the same size as an integer -- to the function, so there is no additional memory allocated to copy the object.
+   * The original variable can be changed by the function, and so any changes that happen within the function are retained after we leave the function's scope. 
+
+We indicate that we want a reference to a variable using the `&` operator. The function below will take a reference to an integer and increment that integer by one. Because we have changed the value stored at that memory location, once we leave this function the variable that we passed to it will retain this increased value. 
+```cpp
+void increment(int &x)
+{
+    x = x + 1;
+}
+```
+Passing by reference can save significant time and memory by avoiding making needless copies of variables, but at the cost of making variables potentially vulnerable to being changed by a function. This makes it harder for someone using the function to reason about the program, and what the value of the variables they pass in will be once the function has finished.  
+
+## Using `const` in Pass By Reference
+
+We can retain the performance advantages of pass by reference and still protect our variables from changes by passing a const reference. 
+```cpp
+void constRefExample(int const &x)
+{
+    return 5*x; 
+}
+```
+The declaration `int const &x` means that `x` is a reference (`&`) to a constant (`const`) integer (`int`). This means that the integer value cannot be changed, and so any attempt to change the value of `x` in the function will lead to a compiler error. 
+
+## Using references for output variables
+
+When we use a `return` statement in a function, we are also passing by value, and a copy of the variable that we are returning is made. Just like with inputs to a function, this can be a performance issue if the data that we want to output is large. 
```

Comment:
I think this should be more precise, i.e. the variable is copied using the object's copy-constructor (I think we can assume students either already know the cop-constructor concept or can look it up).

You should probably also mention return value optimisation as a possible alternative to passing output vars by reference.

Reply:
Agreed. Have added basic discussion of copy constructor and move constructor and how they relate to `return`. Have also changed `vector` to an arbitrary `ImmovableData` class because I realised `vector` _does_ has a move constructor and therefore doesn't have a copy problem when returning.